### ISSUE TITLE
It is not currently possible to create a custom Flow Contract that appears on Rarible.com

### ISSUE THREAD

**Issue by figs999** [rarible/protocol#173]
### Discussed in https://github.com/rarible/protocol/discussions/167

This was originally a discussion, but I am opening an issue to track progress as this needs to be addressed as quickly as possible for my project.

@atapin mentioned that this was currently not possible because of the way that the indexer needs to interact with flow contracts and that the work around being discussed internally involved contract developers hosting an endpoint of some kind that would provide the token metadata.

My suggestion is to add a new contract interface that all compatible contracts would be required to implement that has a contract level public function that can be used to query the metadata for any token in the collection by ID. This function would have identical return type as the getMetadata function defined at resource level in the RaribleNFT contract:

A simple example implementation, assuming all metadata is stored externally in json using the standard format:
```
    pub fun GetMetaDataFor(tokenId: UInt64): {String:String}  {
        return {"metaURI": self.baseURL.concat(tokenId.toString())}
    }
```

**Comment by figs999** [comment]
I've brought the issue to the other members of the NFT metadata standards committee for Flow.

Another member suggested that it may be easier for a wide variety of contracts to implement a standard practice of emitting the metadata json data for each NFT as an event whenever it is altered. 

This would allow projects to maintain resource ownership of the metadata if they wanted to and would allow them to use alternative metadata storage methods (such as in native data structures on chain) if they wanted to.

**Comment by atapin** [comment]
@figs999 can you please link the discussion in the committee? We're taking this topic for work next week. So I think it would be great to have a common approach.

Also, please provide the address of your deployed contract.

**Comment by atapin** [comment]
@figs999 did you deployed the contract? can I help?

**Comment by figs999** [comment]
@atapin No, the contract is not yet deployed. We are still in development.

However, isn’t the solution that was proposed that we need to host an API endpoint that will provide the metadata? I just need the specifications for that endpoint if that is the case.

**Comment by atapin** [comment]
@figs999 we're still testing different approaches. Most likely, we won't go with ad-hoc APIs for the metadata. Let's keep in touch on the progress.

**Comment by eduardstal** [comment]
Closed this due to the thread owner not responding. Reopen if needed. 

https://github.com/rarible/protocol/discussions/167
